### ISSUE TITLE
Renaming mounted FS with child caused apparent loss of child contents

### ISSUE THREAD

**Issue by slburson** [openzfs/zfs#15763]
### System information

Type | Version/Name
 --- | ---
Distribution Name	| openSUSE Leap
Distribution Version	| 15.5
Kernel Version	| 5.14.21-150500.55.39-default
Architecture	| x86_64
OpenZFS Version	| 2.2.2-1

Possibly related to/duplicate of #13911.

### Describe the problem you're observing

I had a dataset `pool/A` mounted on `/A`, with a child `pool/A/B` mounted on `/A/B`.  I did this (from memory, but I think it's right):
```
zfs set mountpoint=/C pool/A
zfs set mountpoint=/C/B pool/A/B
zfs rename -u pool/A pool/C
```
I wound up with datasets `pool/C` and `pool/C/B` as expected, with `pool/C` mounted on `/C` and `pool/C/B` mounted on `/C/B` as expected, but `/C/B` was empty.  :scream_cat:  I thought maybe unmounting and remounting would bring it back, but `zfs unmount pool/C/B` failed with an error.  Fortunately, rebooting cured the problem 😅 — the contents of `/C/B` returned.

### Describe how to reproduce the problem

If the above procedure doesn't do it for you, let me know and I'll try it again myself.

**Comment by rincebrain** [comment]
I mean, it sounds like it unmounted pool/C/B unconditionally then didn't mount it again after the rename, because you used -u.

But that would just be a guess. Can you reproduce this?

**Comment by slburson** [comment]
I don't think it's that simple — `df` reported it as mounted.  But let me try again.

**Comment by rincebrain** [comment]
That doesn't look like a bug.

You just tried to rename something that, after unmounting anything with a mountpoint relative to the parent you're renaming, still had something using the thing, so it couldn't unmount it, because the mount is still in use.

**Comment by slburson** [comment]
Oh, I see.  When there is no explicit mountpoint set for the child, moving the parent moves the child, but when there is, it can't move the child.

Okay, will keep looking...

**Comment by usaleem-ix** [comment]
I think that this is not a bug. I tried to reproduce the first problem related to data loss, but I was not able to do so.

However, I was able to reproduce the other issue mentioned above where unmount errors out with permission denied while trying to update the mountpoint property for dataset A. I think that this also is not a bug.

When we do the following:
```
# zfs set mountpoint=/A pool/A
# zfs set mountpoint=/A/B pool/A/B
```
This updates the `source` for mountpoint property for both child and parent datasets. If we only update the mountpoint property for parent dataset (A), in that case, the child dataset (B) mountpoint is also updated. The `source` for parent dataset is `local` but for child dataset (B), the source for mountpoint property is `Inherited from pool/A`. You can change the mountpoint for parent (A) to whatever desired and the child dataset will be moved along with it, without any issue.

Now when we explicitly set the mountpoint property for child dataset (B), to reside under `/A`, the source for mountpoint property for child dataset (B) is updated from `Inherited` to `Local`. In this scenario when we try to update the mountpoint for parent dataset (A), it cannot unmount because another dataset is mounted under it and it's mountpoint is not inherited. That's why the unmount operation fails while trying to update the mountpoint for dataset (A).

You should be able to update the mountpoint for dataset A, by first inheriting the mountpoint for child dataset (B) by doing `zfs inherit mountpoint pool/A/B`. OR by updating the mountpoint for dataset B to anything outside dataset A. OR unmount B first before updating the mountpoint for A.

**Comment by slburson** [comment]
I now believe I was simply confused.  I think that, as @rincebrain said [above](https://github.com/openzfs/zfs/issues/15763#issuecomment-1888420050), `pool/C/B` was unmounted, and I misinterpreted what `df` was showing me.

Closing as invalid.
